Add exception property to IInvocation

MaStr11 opened the pull request:
Fixes #1070

MaStr11 commented on src/Moq/IInvocation.cs:
```diff
@@ -35,5 +36,13 @@ public interface IInvocation
 		/// Gets the return value of the invocation.
 		/// </summary>
 		object ReturnValue { get; }
```
We may add a `remark` her too:
```suggestion
		/// <remarks>
		/// Returns <see langword="null" /> if the invocation did throw an <see cref="Exception" /> or was 
		/// a <see langword="void" /> method.
		/// </remarks>
		object ReturnValue { get; }
```

MaStr11 replied:
Done 988feb05 (I like the Moq 5 comments!)

MaStr11 commented on src/Moq/Invocation.cs:
```diff
@@ -69,16 +69,25 @@ public MethodInfo MethodImplementation
 
 		public Type ProxyType => this.proxyType;
 
+		/// <inheritdoc cref="IInvocation.ReturnValue"/>
```
I don't know if this is needed/supported and I have not checked whether it is working or not.

MaStr11 commented on src/Moq/ProxyFactories/CastleProxyFactory.cs:
```diff
@@ -101,6 +101,15 @@ public void Intercept(Castle.DynamicProxy.IInvocation underlying)
+				catch (Exception ex)
+				{
+					invocation.ReturnValue = new InvocationExceptionWrapper(ex);
+					throw;
```
I think `throw`ing the original exception is the right thing to do here:
https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/throw#re-throwing-an-exception

MaStr11 commented on src/Moq/ProxyFactories/CastleProxyFactory.cs:
```diff
@@ -101,6 +101,15 @@ public void Intercept(Castle.DynamicProxy.IInvocation underlying)
+				{
+					throw;
+				}
+				catch (Exception ex)
```
Is there a better way to intercept exceptions (I'm not really familiar with *CastleProxy*)?

MaStr11 replied:
I thought the `Expcetion`property is for exceptions thrown by the method. `MockException` is not raised as part of the invocation, but before or after. I can't tell which is more sound: Storing MockException in `Exception` or not. Either way, the behavior should be covered by tests (which it is not yet).

Can you just tell me which way to go here?

MaStr11 replied:
I changed the behavior so that `MockException` is stored in the `Exception` property and added a test in a064c982. Just tell me, if you find that behavior appropriate or not.

MaStr11 reviewed (commented):
@stakx Can you please have a look at my solution? Does this come close to what you had in mind?

stakx commented on tests/Moq.Tests/InvocationsFixture.cs:
```diff
@@ -45,11 +47,109 @@ public void MockInvocationsIncludeArguments()
+			};
+
+			Assert.Throws<ArgumentException>(() => mock.Object.GetRange(1, 1));
+			// Base-call exceptions are not recorded
+			Assert.Empty(mock.Invocations);
```
Good of you for thinking of this test case! (I didn't.)

Something's odd here&mdash;invocations should get recorded even when `mock.CallBase == true`. I think the reason why no invocations get recorded here is that `List<>.GetRange` is not virtual.

Could you change this test so it calls some overridable method?

MaStr11 replied:
Didn't think of the virtual requirement. I will add another test and change the code comment accordingly.

MaStr11 replied:
I added another test 09816d01

stakx reviewed (changes requested):
Thanks for working on this, this is some great work so far! :+1:

I do have a few change requests, mostly regarding details. Apart from the comments below, could you also add an entry to the `CHANGELOG.md` again? (You can also change your previous entry for `ReturnValue` to now include `Exception`, if you'd like. In that case, just make sure to mention both PR numbers at the line's end.)

MaStr11 commented on tests/Moq.Tests/InvocationsFixture.cs:
```diff
@@ -45,11 +47,109 @@ public void MockInvocationsIncludeArguments()
+			var exceptionThrownAndHandled = false;
+			mock.Setup(c => c.CompareTo(It.IsAny<object>())).Throws(exception);
+
+			try
+			{
+				mock.Object.CompareTo(null);
+			}
+			catch (Exception thrown)
+			{
+				Assert.Equal(exception.Message, thrown.Message);
+				var invocation = mock.Invocations[0];
+
+				Assert.Same(thrown, invocation.Exception);
+				exceptionThrownAndHandled = true;
+			}
```
I know, but I need the `Exception thrown` object for the assert (I want to make sure it's the same object, to catch any breaking changes if someone replaces the `throw;` with `throw ex;`)

stakx replied:
`Assert.Throws` returns the caught exception, so you can assign it to a variable.

MaStr11 replied:
That's much better (60550157)!

MaStr11 commented on src/Moq/ProxyFactories/CastleProxyFactory.cs:
```diff
@@ -146,8 +155,8 @@ public override bool ShouldInterceptMethod(Type type, MethodInfo method)
 			private static bool IsRelevantObjectMethod(MethodInfo method)
 			{
 				return method.DeclaringType == typeof(object) && (method.Name == nameof(object.ToString)
-				                                              ||  method.Name == nameof(object.Equals)
-				                                              ||  method.Name == nameof(object.GetHashCode));
+															  || method.Name == nameof(object.Equals)
+															  || method.Name == nameof(object.GetHashCode));
```
Looks good in Visual Studio:
![image](https://user-images.githubusercontent.com/24472128/96036688-1dd64700-0e65-11eb-9855-f2f8e35fc3f1.png)

I think it is a tabs vs spaces issue. I'd like to leave it as is if that's okay for you.

stakx replied:
It only looks good if you've set tabs to equal 4 spaces. Other settings will lead to misalignment.

MaStr11 replied:
Okay. I reverted it.

MaStr11 commented on tests/Moq.Tests/InvocationsFixture.cs:
```diff
@@ -45,11 +47,109 @@ public void MockInvocationsIncludeArguments()
+
+			var invocation = mock.Invocations[0];
+
+			Assert.Equal(534011718, invocation.ReturnValue);
```
I went for the second option 31f5ec7e.

MaStr11 commented:
@stakx I think I addressed all your comments (nice feedback, btw). I think the last open topic is whether `MockExceptions` should be assigned to `Exception` or not: https://github.com/moq/moq4/pull/1077/files#r504932581

stakx commented on tests/Moq.Tests/InvocationsFixture.cs:
```diff
@@ -45,11 +47,131 @@ public void MockInvocationsIncludeArguments()
+
+			var thrown = Assert.Throws<MockException>(() => mock.Object.Clone());
+
+			Assert.Equal(
+@"ICloneable.Clone() invocation failed with mock behavior Strict.
+All invocations on the mock must have a corresponding setup.", thrown.Message);
```
All good. If you wanted to simplify this test code a little further, you could query `thrown.Reasons == MockExceptionReasons.NoSetup` instead of comparing the message. That'll make the test less sensitive to potential future changes to that message's precise wording. (`MockException.Reasons` is an internal property that exists specifically for our test suite.)

stakx reviewed (approved):
Looks great, thanks for making the changes!

I'm happy to merge this PR as it stands, however I do have two more comments (below) if you wish to refine things a little further. That's purely optional, though... otherwise, I'll merge your PR some time tomorrow.

MaStr11 commented:
I addressed your feedback and the PR should be done now. If I should do anything further, just let me know.

stakx commented:
@MaStr11, excellent work, thanks for contributing once again!
